File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int l = 0 , r = 0 , maxlength = 0 , len = s.length();
        int[] arr = new int[256];
        while( r < len ){
            char c = s.charAt(r);
            arr[c] += 1;
            while (arr[c] != 1){
                arr[ s.charAt(l) ] -= 1;
                l++;
            }
            maxlength= Math.max( maxlength , r-l+1 );
            r++;
        }
        return maxlength;
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        l,result = 0,0
        window_state = {}
        for r in range(len(s)):
            char = s[r]
            window_state[char]=window_state.get(char,0)+1

            while window_state[char]>1:
                left_char = s[l]
                window_state[left_char]-=1
                if window_state[left_char] == 0:
                    del window_state[left_char]
                l+=1

            result = max(result,r-l+1)

        return result
        